pymc3/sampling.py

```diff
@@ -580,6 +626,267 @@ def _iter_sample(draws, step, start=None, trace=None, chain=0, tune=None,
             step.report._finalize(strace)
 
 
+class PopulationStepper(object):
```

Comment:
What about the code we have for SMC? Is there any overlap between these?

Reply:
The problem is the same, but the solution is different.  And performance is a concern, so it's worthwhile to think about it.

SMC uses a `multiprocessing.Pool` and a text file `MultiTrace` backend to share the population state and synchronize chains.  Although I have to admit that I don't completely understand how this works in combination with the `Pool`.

I use Pipes to create a communication channel between the processes. That channel is then used to pass the population to the chain, and the update back to the main iterator.

My guess is that the Pipes should in principle be faster. And they certainly involve less code.

Reply:
My concern is code duplication in the pymc3 code-base.

Reply:
Maybe we can refactor SMC at some point @hvasbath @aloctavodia?

Reply:
I will love to help with SMC, "at some point" could be during January :-)

Reply:
The algorithms themselves are quite different. I am not sure if SMC can be refactored to support the parallel sampling here, which is the only thing the samplers have in common appart from the metropolis step ;). For sure the pool also supports python2.7, which is apparently not supported in the DEM here.
The chains in the SMC do not communicate with each other during the sampling as they do here. They communicate all together every n_chains * nsteps samples. For sure would be interesting to see the test and performance you did also compared to the SMC @michaelosthege . There is the SMC2gaussians notebook where you see how to start it. Anyways if anyone has a good idea to anyway factor out common things I also could help starting earliest January.